sdk/eventhub/event-hubs/src/loadBalancerStrategies/loadBalancingStrategy.ts

```diff
@@ -0,0 +1,398 @@
+ * @ignore
+ * @internal
+ */
+function calculateBalancedLoadCounts(
+  ownerToOwnershipMap: Map<string, PartitionOwnership[]>,
+  partitionIds: string[]
+): { minPartitionsPerOwner: number; requiredNumberOfOwnersWithExtraPartition: number } {
+  // Calculate the minimum number of partitions every EventProcessor should own when the load
+  // is evenly distributed.
+  const minPartitionsPerOwner = Math.floor(partitionIds.length / ownerToOwnershipMap.size);
+
+  // If the number of partitions in the Event Hub is not evenly divisible by the number of active
+  // EventProcesrrors, some EventProcessors may own 1 partition in addition to the minimum when the
+  // load is balanced.
+  // Calculate the number of EventProcessors that can own an additional partition.
+  const requiredNumberOfOwnersWithExtraPartition = partitionIds.length % ownerToOwnershipMap.size;
+
+  return {
+    minPartitionsPerOwner,
+    requiredNumberOfOwnersWithExtraPartition
+  };
+}
+
+/**
+ * Counts the EventProcessors and tallies them by type.
+ *
+ * To be in balance we need to make sure that each EventProcessor is only consuming
+ * their fair share.
+ *
+ * When the partitions are divvied up we will sometimes end up with some EventProcessors
+ * that will have 1 more partition than others.
+ * This can happen if the number of partitions is not evenly divisible by the number of EventProcessors.
+ *
+ * So this function largely exists to support isLoadBalanced() and
+ * shouldOwnMorePartitions(), both of which depend on knowing if our current list
+ * of EventProcessors is actually in the proper state.
+ *
+ * @param minPartitionsPerOwner The number of required partitions per EventProcessor.
+ * @param ownerToOwnershipMap The current ownerships for partitions.
+ * @internal
+ * @ignore
+ */
+function getEventProcessorCounts(
+  minPartitionsPerOwner: number,
+  ownerToOwnershipMap: Map<string, PartitionOwnership[]>
+): EventProcessorCounts {
+  const counts: EventProcessorCounts = {
+    haveRequiredPartitions: 0,
+    haveAdditionalPartition: 0,
+    haveTooManyPartitions: 0
+  };
+
+  for (const ownershipList of ownerToOwnershipMap.values()) {
+    const numberOfPartitions = ownershipList.length;
+
+    // there are basically three kinds of partition counts
+    // for a processor:
+
+    // 1. Has _exactly_ the required number of partitions
+    if (numberOfPartitions === minPartitionsPerOwner) {
+      counts.haveRequiredPartitions++;
+    }
+
+    // 2. Has the required number plus one extra (correct in cases)
+    // where the # of partitions is not evenly divisible by the
+    // number of processors.
+    if (numberOfPartitions === minPartitionsPerOwner + 1) {
+      counts.haveAdditionalPartition++;
+    }
+
+    // 3. has more than the possible # of partitions required
+    if (numberOfPartitions > minPartitionsPerOwner + 1) {
+      counts.haveTooManyPartitions++;
+    }
+  }
+
+  return counts;
```

Comment:
I had to check (and I think this is true) that we don't care about any processors that fall into none of those groups (ie, ones that don't own even the minPartitionsPerOwner).

Reply:
Correct. Since we know how many processors there are, we can infer how many don't fall into any of these groups if that's needed.